step-71: the biharmonic equation.

bangerth opened the pull request:
I'd like to get this merged sooner rather than later because we've got several people who want to work on finishing this program together.

bangerth commented:
/rebuild

tjhei commented:
can you please rebase to master? This is needed to make the testers happy.

rumajo commented on examples/step-71/doc/intro.dox:
```diff
@@ -0,0 +1,159 @@
+first derivatives. This is the realm of elements such as the Argyris
+element, the Clough-Tocher element and others, all developed in the
+late 1960s. From a twenty-first century perspective, they can only be
+described as bizarre in their construction. They are also ecxeedingly
```
ecxeedingly -> exceedingly

bangerth replied:
Thanks for pointing it out :-)

bangerth commented:
Rebased!

bangerth commented on examples/step-71/step-71.cc:
```diff
@@ -0,0 +1,720 @@
+      const auto &q_points = fe_i.get_quadrature_points();
+
+      copy_data.face_data.emplace_back();
+      CopyDataFace &copy_data_face = copy_data.face_data.back();
```
@tjhei -- you're adding a face data object to the end of the vector here (and you do the same for the boundary worker), and then their contents are copied in the `copier` function. But you never actually remove them from the vector again. So I have a suspicion that you're actually copying the same data over and over into the matrix. Do you agree that this looks like a bug?

tjhei commented on examples/step-71/doc/intro.dox:
```diff
@@ -0,0 +1,461 @@
+$K_{+},K_{-} \in \mathbb{T}$), we cope with this discontinuity by
+defining the following single-valued functions on $e$:
+@f{align*}{
+\Bigg[\!\Bigg[\frac{\partial^k v_h}{\partial \mathbf n^k}\Bigg]\!\Bigg]&= \frac{\partial^k v_h|_{K_+}}{\partial \mathbf n^k} \bigg |_e - \frac{\partial^k v_h|_{K_-}}{\partial \mathbf n^k} \bigg |_e,\\
```
we use [u] and {u} in FEInterfaceValues, see https://www.dealii.org/developer/doxygen/deal.II/classFEInterfaceValues.html#a7827502cb506dbd5d6b2f740184f0a21

do you mind switching to that notation?

tjhei commented on examples/step-71/doc/results.dox:
```diff
@@ -0,0 +1,167 @@
+  </tr>
+</table>
+In the case, all convergence rates of $u$
+follow the theoretical expectations.
```
I don't understand. We don't see a rate of 2 for L2, which we expect.

tjhei replied:
I would guess this is a suboptimal parameter choice and we should say so (or delete this section).

tjhei commented on examples/step-71/doc/results.dox:
```diff
@@ -0,0 +1,167 @@
+should then be around 2, 2, 1 for $Q_2$; 4, 3, 2 for
+$Q_3$; and 5, 4, 3 for $Q_4$, respectively.
+
+From the papers by Brenner et al., it is not immediately clear what
```
This is what I found:
- Brenner/Gudi/Sung: ">=1", "we take the penalty parameter \sigma to be 5"
- Fenics/dolfin tutorial: "alpha=8"

tjhei replied:
and Wells/Dung 2007 for Kirchhoff plates: stability with " > n_e (number of edges belonging to an element)", section 4.2

tjhei commented:
I left several comments about problems we should think about. @bangerth brought up the question of "why 2 \gamma for the boundary term?". My thoughts about that:
The factor 2 is not needed and not used in Engel/Garikipati/et al, 2002 (they choose tau=tau_h).

bangerth commented:
Thanks, @tjhei. I've done everything here except for the double brackets, which I'll leave for a later time (or maybe change it in the documentation of `FEInterfaceValues`).

bangerth commented:
@dealii/dealii So this then is our Christmas present: A finished step-71. Review away :-)

drwells commented:
:christmas_tree:

drwells commented:
I will take a look at this early this week.

bangerth commented:
Much appreciated, thank you in advance!

peterrum commented on examples/step-71/doc/intro.dox:
```diff
@@ -0,0 +1,488 @@
+&+ \sum_{e \in \mathbb{F}} 
+\frac{\gamma}{h_e}\int_e \bigg[\!\bigg[\! \frac{\partial v_h}{\partial \mathbf n} \bigg]\!\bigg] 
+\bigg[\!\bigg[ \frac{\partial u_h}{\partial \mathbf n} \bigg]\!\bigg] \ ds,
+@f}
```
The second and third term seem to be identical. According to the program, you should swap u and v in the third term.

bangerth replied:
Good catch!

peterrum commented on examples/step-71/doc/intro.dox:
```diff
@@ -0,0 +1,488 @@
+@f{align*}{
+\mathcal{F}(v_h)&:=\sum_{K \in \mathbb{T}}\int_{K} v_h f \ dx +\sum_{e \in \mathbb{F}^b} \frac{\gamma}{h_e}\int_e 
+\frac{\partial v_h}{\partial \mathbf n} j \ ds.
+@f}
```
The second term is only valid on the boundary faces, isn't it?

bangerth replied:
Yes, that's what `\mathbb{F}^b` was supposed to indicate. But I'm making this clearer now.

peterrum commented on examples/step-71/step-71.cc:
```diff
@@ -0,0 +1,937 @@
+                   * jump_grad_i_dot_n             // [grad v n]
+                   * (exact_gradients[qpoint] * n) // (grad u_exact . n)
+                 ) *
+                JxW[qpoint]; // dx
```
Where does the first term come from?

bangerth replied:
Hm, good question. @tjhei: Did we simply forget it in the description in the introduction, where we define the right hand side vector? I think so.

peterrum commented on examples/step-71/step-71.cc:
```diff
@@ -0,0 +1,937 @@
+    // on them, and then the copier function that puts things into the
+    // global matrix and right hand side. As an additional benefit,
+    // MeshWorker::mesh_loop() does all of this in parallel, using
+    // as many processor cores as your machine happens to have.
```
If the data structures (vector and matrix) are set the right way...

bangerth replied:
Sure, but that's not necessary to say, is it?

bangerth commented on examples/step-71/step-71.cc:
```diff
@@ -0,0 +1,937 @@
+    system_matrix.reinit(sparsity_pattern);
+
+    solution.reinit(dof_handler.n_dofs());
+    system_rhs.reinit(dof_handler.n_dofs());
```
Nope :-) One issue per tutorial program, and here the issue is how to solve the biharmonic equation, not the biharmonic equation in parallel :-)

peterrum replied:
As far a I see it, an extension to a parallel version would be like an additional max. 10 lines and wouldn't add really additional complexity (MPI is since recently a basic technique). But I guess I cannot convince anyone...

bangerth commented on examples/step-71/step-71.cc:
```diff
@@ -0,0 +1,937 @@
+    ScratchData(const Mapping<dim> &      mapping,
+                const FiniteElement<dim> &fe,
+                const unsigned int        quadrature_degree,
+                const UpdateFlags         update_flags = update_values |
```
Good question -- a leftover from the program I got from @tjhei :-) Fixed.

bangerth commented on examples/step-71/step-71.cc:
```diff
@@ -0,0 +1,937 @@
+      scratch_data.fe_values.reinit(cell);
+      cell->get_dof_indices(copy_data.local_dof_indices);
+
+      const FEValues<dim> &fe_values = scratch_data.fe_values;
```
Good catch.

bangerth commented on examples/step-71/step-71.cc:
```diff
@@ -0,0 +1,937 @@
+    // `copy_data.face_data()` is really all that the later `copier` function
+    // gets to see when it copies the contributions of each cell to the global
+    // matrix and right hand side objects.
+    auto face_worker = [&](const Iterator &    cell,
```
Good idea!

bangerth commented:
@peterrum -- thanks for the feedback! I've incorporated it into the current version with the only exception of the issue with the missing term in the description.

I've also fixed a variable name: `gamma` should really have been `gamma_over_h`.

drwells commented on examples/step-71/doc/intro.dox:
```diff
@@ -0,0 +1,508 @@
+  \int_K v_h (\Delta^2 w_h)
+  &= \int_K (\nabla \nabla v_h) : (\nabla \nabla w_h)
+     +\int_{\partial K} v_h (\nabla \Delta w_h \cdot \mathbf n)
+     -\int_{\partial K} (\nabla v_h \otimes \mathbf n) : (\nabla\nabla w_h)
```
I might be being dense but I'm not sure what `otimes` is supposed to mean here. We define `:` below so I think we should also define this operator.

bangerth replied:
It's the outer product. I fixed that.

drwells commented on examples/step-71/doc/intro.dox:
```diff
@@ -0,0 +1,508 @@
+would make for a good exercise. Instead, we'll simply check in the
+program what the "broken" $H^2$ seminorm
+@f{align*}{
+ |u_h|^\circ_{H^2}^2
```
We need some extra braces here to clarify the exponentiation (my copy of mathjax complains)

drwells commented on examples/step-71/doc/intro.dox:
```diff
@@ -0,0 +1,508 @@
+convergence orders.
+
+For the purposes of this program, we're a bit too lazy to actually
+implement this equivalent norm -- though it's not very difficult and
```
```suggestion
implement this equivalent seminorm -- though it's not very difficult and
```

drwells commented on examples/step-71/step-71.cc:
```diff
@@ -0,0 +1,955 @@
+// The two most interesting header files will be these two:
+#include <deal.II/fe/fe_interface_values.h>
+#include <deal.II/meshworker/mesh_loop.h>
+// The first of these is responsible for providing the class FEInterfaceValue
```
```suggestion
// The first of these is responsible for providing the class FEInterfaceValues
```

bangerth commented on doc/doxygen/tutorial/tutorial.h.in:
```diff
@@ -950,7 +970,7 @@
  *
  *   <tr valign="top">
- *     <td> The heat equation
+ *     <td> Heat equation
```
I think we're not consistent in this file; the previous and next few items in the list started here are:
* Helmholtz equation
* Elasticity and elasto-plasticity equations
* Minimal surface equation
* Quasi-stationary elasticity equation.

I thought removing the "The" in this one place would be useful. Let's leave making it better for a later PR.

bangerth commented on examples/step-71/doc/intro.dox:
```diff
@@ -0,0 +1,488 @@
+  &&\forall \mathbf x \in \partial\Omega,
+@f}
+then the following trick works (at least if the domain is convex, see
+below): In much the same as we obtained the
```
Is your comment independent of the change request? If yes, can you clarify what's unclear to you?

drwells replied:
I should have been clearer: I just didn't understand the fragment "In much the same we obtained", it sounds like its missing a word to me. The suggested change fixes my issue.

drwells replied:
I should have been clearer. The original version of that sentence was confusing to me because it looked like it was missing a word. The corrected version makes sense to me.

bangerth commented on examples/step-71/doc/intro.dox:
```diff
@@ -0,0 +1,488 @@
+  than just $H^1(\Omega)$. This can be argued as follows: For convex
+  domains,
+  <a href="https://en.wikipedia.org/wiki/Elliptic_operator#Elliptic_regularity_theorem">"elliptic
+  regularity"</a> implies that if the right hand side $v\in H^s$, then
```
Well, smoothly bounded domains are boring -- the world has corners. I've added a couple of sentence to the comment.

drwells replied:
Ah, I see - that goes with the corners. Makes sense now!

bangerth commented on examples/step-71/doc/intro.dox:
```diff
@@ -0,0 +1,508 @@
+domain and then integrating by parts can't work.
+
+Historically, numerical analysts have tried to address this by
+inventing finite elements that are "C<sup>1</sup> continuous", i.e., that use
```
I already fixed that in response to the review by @peterrum .

drwells replied:
:+1:

bangerth commented on examples/step-71/doc/intro.dox:
```diff
@@ -0,0 +1,488 @@
+\bigg[\!\bigg[ \frac{\partial u_h}{\partial \mathbf n} \bigg]\!\bigg].
+@f}
+Then, after making cancellations that arise, we arrive at the following $C^0$
+IP formulation of the biharmonic equation: find $u_h$ such that $u_h =
```
As you wish :-)

bangerth commented on examples/step-71/doc/intro.dox:
```diff
@@ -0,0 +1,488 @@
+@f}
+yields. The convergence rate in this norm can, from a theoretical
+perspective, of course not be <i>worse</i> than the one for
+$|\cdot|_h$, but it could be worse. It could also be the case that
```
It could be better. Fixed.

bangerth commented on examples/step-71/doc/results.dox:
```diff
@@ -0,0 +1,211 @@
+the penalty parameter $\gamma$ should be. For example,
+@cite Brenner2009 state that it needs to be larger than one, and
+choose $\gamma=5$. The FEniCS/Dolphin tutorial chooses it as
+$\gamma=8$. @cite Wells2007 uses a value for $\gamma$ larger than the
```
Yes, we can!

bangerth commented:
Thanks, @drwells -- I've applied all of your comments.

drwells commented:
That was quick! I am taking a second look now :)

drwells commented on examples/step-71/step-71.cc:
```diff
@@ -0,0 +1,952 @@
+  // @sect4{Assembling the linear system}
+  //
+  // The following pieces of code are more interesting. They all relate to the
+  // assembly of the linear system. While assemling the cell-interior terms
```
```suggestion
  // assembly of the linear system. While assembling the cell-interior terms
```

drwells commented on examples/step-71/step-71.cc:
```diff
@@ -0,0 +1,952 @@
+      // indices `i` and `j` to add up the contributions of this face
+      // or sub-face. These are then stored in the
+      // `copy_data.face_data` object created above. As for the cell
+      // worker, we pull the evalation of averages and jumps out of
```
```suggestion
      // worker, we pull the evaluation of averages and jumps out of
```

drwells commented:
I think this is good to go once @peterrum's comments are addressed.

bangerth commented:
@tjhei: Any insight into the missing term? I think that it's really just missing in the introduction.

bangerth commented:
I'll make a different offer: If someone merges, then I'll open an issue that tracks the confusion with the missing term. I'm a bit antsy to get this merged and off my plate.

drwells commented on examples/step-71/step-71.cc:
```diff
@@ -0,0 +1,952 @@
+      copy_data.cell_matrix = 0;
+      copy_data.cell_rhs    = 0;
+
+      const FEValues<dim> &fe_values = scratch_data.fe_values;
```
```suggestion
      FEValues<dim> &fe_values = scratch_data.fe_values;
```
I should have caught this before!

drwells commented:
@peterrum the remaining unaddressed comments are for improvements and the one thing mentioned by @bangerth that needs to check the missing term. Are you okay with merging it in its present state?

peterrum commented:
Let's pass the torch to @tjhei!

tjhei commented:
> Let's pass the torch to @tjhei!

I will need to go through the formulas and derivation again. I wouldn't mind merging and figuring this out later.

drwells commented:
Since @bangerth indicated there is some urgency in merging this, lets merge this as-is and open an issue to verify things before 9.2. I think we have all agreed on this.

bangerth commented:
Yay!
